Floats can be created from numpy floats

cbm755 opened the pull request:
This is rebasing the work of @bmgxyz and @Trave11er who both worked on the same issue #13275.

Also, update the AUTHORS file.  @bmgxyz and @Trave11er please confirm that this is how you would like your name/email/etc.

I'm not planning on merging this until all involved are happy.  And I can rebase it if there is further changes at #13400 and #13401.

bmgxyz committed:
sympify: Made it so that basic numpy datatypes can be cast as Float

Factored out changes made in #11666 into a separate function called
`convert_numpy_types` in sympify.py. Also called that function both in
sympify.py and in Float() when the input is a numpy type. If the input is a
numpy type that can't be cast as a Float, it passes through to the rest of
sympify. Fixes #13275.

Previously, the following would fail:

>>> Float(np.float128(5))
>>> Float(np.float32(5))

but now they don't.

Trave11er committed:
test_sympify: reorder, skip immediately if import fails

Trave11er committed:
Float: add tests for numpy conversion

cbm755 committed:
Float: minor edits to numpy tests

Test a repeating decimal instead of integer.  Ensure abs() includes
the denominator in relative error.  Weaken some tests a bit.

cbm755 committed:
Add new contributor Bradley Gannon

Welcome to SymPy!

cbm755 committed:
Add new contributor Gleb Siroki

Welcome to Sympy!

bmgxyz reviewed (approved):
Looks good to me. If we end up needing doctests for ```_convert_numpy_types()```, then we can deal with that in the future.

@cbm755 Thanks for taking the time to combine our work. :+1:

Trave11er commented:
Indeed - thank you, @cbm755 for reviewing our code and putting it all together.

One small thing - in test_numbers.py the function name is `check_prec_and_relerr` but the call is for `check_precision_and_relerr`. Otherwise, looks good.

cbm755 commented:
Nice catch @Trave11er; I've fixed and rebased. 

So then why did the tests pass?  Travis never runs with numpy apparently.

cbm755 commented:
@asmeurer I'm a bit too close to this now, can you review these tests?  I'm +1 for the other changes.

cbm755 committed:
Float: also test float16 for numpy conversion

ylemkimon commented:
> So then why did the tests pass? Travis never runs with numpy apparently.

Travis runs only selected tests with NumPy. Could you add `sympy/core/tests/test_numbers.py` to it?
https://github.com/sympy/sympy/blob/0cfd9a739c9b8095cbd8d42fe154af4634c7841d/bin/test_travis.sh#L80-L83

asmeurer commented:
Looks good. Can someone double check that the tests pass on 32-bit?

ylemkimon commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,32 @@ def test_Integer_precision():
+    check_prec_and_relerr(np.float128(2)/3, S(2)/3)
+
+    # float128 more precise than double (how much more is arch dep)
+    x = np.float128(2)/3
```
`float128` may not be available exist on all platforms. You should do attribute check or remove it.
Fails on Windows 32-bit: [Appveyor](https://ci.appveyor.com/project/ylemkimon/sympy/build/1.0.16).

cbm755 replied:
Thanks for this!  [The docs](https://docs.scipy.org/doc/numpy/user/basics.types.html) say:

>  In spite of the names, np.float96 and np.float128 provide only as much precision
>  as np.longdouble, that is, 80 bits on most x86 machines and 64 bits in standard Windows builds.

So yes this particular test should be removed.

However, the docs also say:

>  np.float96 and np.float128 are provided for users who want specific padding.

But you say `float128` may not be available.  And indeed `np.float96` is not available on my GNU/Linux 64-bit!  I guess I'll change the test to `longdouble`.

ylemkimon commented:
@asmeurer I think there should be a Windows testing, e.g., AppVeyor, for at least optional dependencies(native CPython extensions). During testing of above PR, I found sympify test fails on Windows 64-bit(#13454).

asmeurer commented:
Agreed

cbm755 committed:
CI: add test_numbers to tests run with numpy

cbm755 committed:
Float: use longdouble instead of float128

It seems "float128" is not portable.  Longdouble may be simply
double precision (e.g., on MSVC compiler on Windows), so stop
testing that is is more accurate.  Note: if longdouble is 80 bits
(as is common) then it is still tests that `Float(<longdouble)>)`
has the correct precision.

cbm755 committed:
test_sympify: rename test and skip, numpy already loaded

cbm755 commented:
@ylemkimon thanks for the feedback, I think I've addressed your review.  Ok to merge this?
